vsr/docker/launch.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+
+import datetime
+import logging
+import os
+import re
+import signal
+import sys
+import telnetlib
+import time
+
+import vrnetlab
+
+def handle_SIGCHLD(signal, frame):
+    os.waitpid(-1, os.WNOHANG)
+
+def handle_SIGTERM(signal, frame):
+    sys.exit(0)
+
+signal.signal(signal.SIGINT, handle_SIGTERM)
+signal.signal(signal.SIGTERM, handle_SIGTERM)
+signal.signal(signal.SIGCHLD, handle_SIGCHLD)
+
+TRACE_LEVEL_NUM = 9
+logging.addLevelName(TRACE_LEVEL_NUM, "TRACE")
+def trace(self, message, *args, **kws):
+    # Yes, logger takes its '*args' as 'args'.
+    if self.isEnabledFor(TRACE_LEVEL_NUM):
+        self._log(TRACE_LEVEL_NUM, message, args, **kws)
+logging.Logger.trace = trace
+
+class VSR_vm(vrnetlab.VM):
+    def __init__(self, username, password):
+        for e in os.listdir("/"):
+            if re.search(".qcow2$", e):
+                disk_image = "/" + e
+        super(VSR_vm, self).__init__(username, password, disk_image=disk_image, ram=1024)
+        self.qemu_args.extend(["-boot", "n", "-monitor", "tcp:0.0.0.0:6000,server,nowait"])
+        self.num_nics = 4
+
+    def bootstrap_spin(self):
+        """ This function should be called periodically to do work.
+        """
+
+        if self.spins > 300:
+            # too many spins with no result ->  give up
+            self.stop()
+            self.start()
+            return
+
+        (ridx, match, res) = self.tn.expect([b"Performing automatic"], 1)
+        if match: # got a match!
+            if ridx == 0: # login
+                self.logger.debug("VM started")
+
+                self.logger.debug("Connecting to QEMU Monitor")
+                self.tn = telnetlib.Telnet("127.0.0.1", 6000 + self.num)
+                self.wait_write("", wait=")")
+
+                # To allow access to aux0 serial console
+                self.logger.debug("Writing to QEMU Monitor")
+
+                # Cred to @plajjan for this one
+                commands = """\x04
+
```

Comment:
The indent you've used here for the multi-line value using `""" ... """` is going to be literal so you are going to be finding a lot of spaces in the output (or rather "sendkey spc", since we are doing that translation). Did you try that it actually works? I guess the CLI doesn't really care about extra spaces so could be fine but is technically suboptimal.

Do indent like:
```
                          commands = """\x04
bla bla
bla bla
bla bla
"""
```